src/visor/visor.go

```diff
@@ -339,210 +409,390 @@ func (vs *Visor) SignBlock(b coin.Block) coin.SignedBlock {
+				return err
+			}
+
+			if bk == nil {
+				return fmt.Errorf("No block exists in depth: %d", tx.BlockSeq)
+			}
+
+			txns = append(txns, Transaction{
+				Txn:    tx.Tx,
+				Status: NewConfirmedTransactionStatus(h, tx.BlockSeq),
+				Time:   bk.Time(),
+			})
 		}
 
-		if bk == nil {
-			return []Transaction{}, fmt.Errorf("No block exsit in depth:%d", tx.BlockSeq)
+		// Look in the unconfirmed pool
+		uxs, err := vs.Unconfirmed.GetUnspentsOfAddr(tx, a)
+		if err != nil {
+			return err
 		}
 
-		txns = append(txns, Transaction{
-			Txn:    tx.Tx,
-			Status: NewConfirmedTransactionStatus(h, tx.BlockSeq),
-			Time:   bk.Time(),
-		})
-	}
+		for _, ux := range uxs {
+			utxn, err := vs.Unconfirmed.Get(tx, ux.Body.SrcTransaction)
+			if err != nil {
+				return err
+			}
+
+			if utxn == nil {
+				logger.Critical("Unconfirmed unspent missing unconfirmed txn")
+				continue
+			}
 
-	// Look in the unconfirmed pool
-	uxs := vs.Unconfirmed.GetUnspentsOfAddr(a)
-	for _, ux := range uxs {
-		tx, ok := vs.Unconfirmed.Get(ux.Body.SrcTransaction)
-		if !ok {
-			logger.Critical("Unconfirmed unspent missing unconfirmed txn")
-			continue
+			txns = append(txns, Transaction{
+				Txn:    utxn.Txn,
+				Status: NewUnconfirmedTransactionStatus(),
+				Time:   uint64(nanoToTime(utxn.Received).Unix()),
+			})
 		}
-		txns = append(txns, Transaction{
-			Txn:    tx.Txn,
-			Status: NewUnconfirmedTransactionStatus(),
-			Time:   uint64(nanoToTime(tx.Received).Unix()),
-		})
+
+		return nil
+	}); err != nil {
+		return nil, err
 	}
 
 	return txns, nil
 }
 
 // GetTransaction returns a Transaction by hash.
 func (vs *Visor) GetTransaction(txHash cipher.SHA256) (*Transaction, error) {
-	// Look in the unconfirmed pool
-	tx, ok := vs.Unconfirmed.Get(txHash)
-	if ok {
-		return &Transaction{
-			Txn:    tx.Txn,
-			Status: NewUnconfirmedTransactionStatus(),
-			Time:   uint64(nanoToTime(tx.Received).Unix()),
-		}, nil
-	}
+	var txn *Transaction
 
-	txn, err := vs.history.GetTransaction(txHash)
-	if err != nil {
-		return nil, err
-	}
+	if err := vs.DB.View(func(tx *bolt.Tx) error {
+		// Look in the unconfirmed pool
+		utxn, err := vs.Unconfirmed.Get(tx, txHash)
+		if err != nil {
+			return err
+		}
 
-	if txn == nil {
-		return nil, nil
-	}
+		if utxn != nil {
+			txn = &Transaction{
+				Txn:    utxn.Txn,
+				Status: NewUnconfirmedTransactionStatus(),
+				Time:   uint64(nanoToTime(utxn.Received).Unix()),
+			}
+			return nil
+		}
 
-	headSeq := vs.HeadBkSeq()
+		htxn, err := vs.history.GetTransaction(tx, txHash)
+		if err != nil {
+			return err
+		}
 
-	confirms := headSeq - txn.BlockSeq + 1
-	b, err := vs.GetBlockBySeq(txn.BlockSeq)
-	if err != nil {
-		return nil, err
-	}
+		if htxn == nil {
+			return nil
+		}
+
+		headSeq, ok, err := vs.Blockchain.HeadSeq(tx)
+		if err != nil {
+			return err
+		} else if !ok {
+			return errors.New("Blockchain is empty but history has transactions")
+		}
+
+		b, err := vs.GetSignedBlockBySeq(htxn.BlockSeq)
```

Comment:
is it ok to call `vs.GetSignedBlockBySeq` in `vs.DB.View`?, the `vs.GetSignedBlockBySeq` method will call `vs.DB.View` as well.

Reply:
It might work if it uses an RLock but still, not ok, inside these visor methods it should call the storage's`*tx` method and not another top-level visor method